list of basic rich text field options

danielfmiranda opened the pull request:
Closes #7756 


The purpose of this PR is to have a uniform list of "base" fields for rich text that we can use for the majority of rich text fields in the CMS.

If a user requests a custom list of features, we can either enter a list of those features if they do not include "bold, italic, and link" ( for example the block quotes can only use bold), or, we can append a list of additional features that they want after base_rich_text_fields.

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/recent_blog_entries.py:
```diff
@@ -3,10 +3,14 @@
 from networkapi.wagtailpages.utils import get_locale_from_request
 
 
+def get_blog_page_categories():
+    from ..blog.blog_category import BlogPageCategory  # noqa
+    return BlogPageCategory.get_categories
+
+
```
why would this be required? (the categories don't

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/base.py:
```diff
@@ -175,9 +176,7 @@ class ParticipatePage2(PrimaryPage):
-        features=[
-            'bold', 'italic', 'link',
-        ],
+        features=(base_rich_text_options),
```
Putting `()` around things is a dangerous thing in Python - at best it does nothing, but often it completely changes the data type, so let's not add them here =)

danielfmiranda replied:
Added to latest commit! Marking as resolved

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_category.py:
```diff
@@ -51,16 +51,21 @@ def get_categories():
+        try:
+            choices = [(cat.name, cat.name) for cat in BlogPageCategory.objects.all()]
+            choices.sort(key=lambda c: c[1])
+        except Exception as err:
+            print(type(err))
+            pass
```
```suggestion
        except ProgrammingError:
            pass
```

danielfmiranda replied:
Added to latest commit! Marking as resolved

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_category.py:
```diff
@@ -6,6 +6,8 @@
 from wagtail.snippets.models import register_snippet
 
+from networkapi.wagtailpages.pagemodels.customblocks.base_rich_text_options import base_rich_text_options
+
```
```suggestion
from django.db.utils import ProgrammingError
```
Although we probably want this directly under the other db import

danielfmiranda replied:
Added to latest commit! Marking as resolved

danielfmiranda replied:
Added to latest commit! Marking as resolved

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/article_fields.py:
```diff
@@ -13,7 +14,7 @@
     ('card_grid', customblocks.CardGridBlock()),
     ('content', customblocks.ArticleRichText(
-        features=['bold', 'italic', 'h2', 'h3', 'h4', 'h5', 'link', 'large', 'ol', 'ul', 'image', 'hr', 'footnotes'],
+        features=base_rich_text_options + ['h2', 'h3', 'h4', 'h5', 'large', 'ol', 'ul', 'image', 'hr', 'footnotes'],
```
does it make sense to unify this with the blog's list as one shared set that is `full_content_rich_text_options = base_rich_text_options + ['h2', 'h3', 'h4', 'h5', 'ol', 'ul', 'hr',]` and then have blog use that directly, with article using `full_content_rich_text_options + ['large', 'image', 'footnotes']`?

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/image_text_block.py:
```diff
@@ -1,10 +1,11 @@
 class ImageTextBlock(ImageBlock):
     text = blocks.RichTextBlock(
-        features=['bold', 'italic', 'h2', 'h3', 'h4', 'h5', 'h6', 'ol', 'ul', 'link']
+        features=base_rich_text_options + ['h2', 'h3', 'h4', 'h5', 'h6', 'ol', 'ul']
```
I suspect we don't need h6 here at all (no on should ever need h6), so we might be about to unify this with the "Expanded"blog/article list too?

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/youtube.py:
```diff
@@ -30,11 +31,10 @@ class YoutubeRegretsPage(FoundationMetadataPageMixin, Page):
             ('paragraph', blocks.RichTextBlock(
-                features=[
-                    'bold', 'italic',
+                features=base_rich_text_options + [
```
this also feels pretty close to blog/article

danielfmiranda commented:
Hi @Pomax, thanks for making some great suggestions regarding a new list with the full functionality of rich text fields. I have added that to this PR and updated the related pages. Thanks again!

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/base_fields.py:
```diff
@@ -9,12 +10,7 @@
-            'ol', 'ul',
-            'link', 'hr',
-        ],
+        features=base_rich_text_options + ['large', 'h2', 'h3', 'h4', 'h5', 'ol', 'ul', 'hr'],
```
Looking at

```
full_content_rich_text_options = base_rich_text_options + ['h2', 'h3', 'h4', 'h5', 'ol', 'ul', 'hr']
```

this looks like it should be `features=full_content_rich_text_options  + ['large']`?

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/recent_blog_entries.py:
```diff
@@ -40,6 +40,8 @@ class RecentBlogEntries(blocks.StructBlock):
     #       error indication what's wrong.
 
     def get_context(self, value, parent_context=None):
+        from ..blog.blog_category import BlogPageCategory  # noqa
```
This line should not be necessary here, given that it's already on line 7 as well.

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_category.py:
```diff
@@ -51,16 +52,20 @@ def get_categories():
+        try:
+            choices = [(cat.name, cat.name) for cat in BlogPageCategory.objects.all()]
+            choices.sort(key=lambda c: c[1])
+        except ProgrammingError:
+            pass
```
Let's add some comments to explain to future readers why this exception is here.

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/customblocks/recent_blog_entries.py:
```diff
@@ -40,6 +40,7 @@ class RecentBlogEntries(blocks.StructBlock):
     #       error indication what's wrong.
 
     def get_context(self, value, parent_context=None):
+
```
```suggestion
```
